src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs

```diff
@@ -23,14 +23,23 @@ class JsonMessageSerializer : IMessageSerializer
             Func<Stream, JsonReader> readerCreator,
             Func<Stream, JsonWriter> writerCreator,
             JsonSerializerSettings settings,
-            string contentType)
+            string contentType,
+            TypeNameHandling typeNameHandling = TypeNameHandling.None)
         {
             this.messageMapper = messageMapper;
 
             if (settings == null)
             {
-                settings = new JsonSerializerSettings { TypeNameHandling = TypeNameHandling.Auto };
-                log.Warn($"The default {nameof(JsonSerializerSettings)} for NServiceBus.Newtonsoft.Json use TypeNameHandling.Auto for backwards compatibility. This is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible. To disable this warning, provide a custom {nameof(JsonSerializerSettings)} instance to 'endpointConfiguration.UseSerialization<NewtonsoftSerializer>().Settings'. Refer to the Json.NET serializer documentation at https://docs.particular.net/ for further details.");
+                settings = new JsonSerializerSettings { TypeNameHandling = typeNameHandling };
+                if (typeNameHandling == TypeNameHandling.Auto)
+                {
+                    log.Warn($"The default {nameof(JsonSerializerSettings)} for NServiceBus.Newtonsoft.Json use TypeNameHandling.Auto for backwards compatibility. This is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible. To disable this warning, provide a custom {nameof(JsonSerializerSettings)} instance to 'endpointConfiguration.UseSerialization<NewtonsoftSerializer>().Settings'. Refer to the Json.NET serializer documentation at https://docs.particular.net/ for further details.");
+                }
+            }
+
+            if (settings.TypeNameHandling == TypeNameHandling.Auto && typeNameHandling != TypeNameHandling.Auto)
+            {
+                log.Warn($"Use of TypeNameHandling.Auto is a potential security vulnerability and it is recommended to use TypeNameHandling.None if possible");
```

Comment:
I think if a user is using the old definition and they don't provide settings they will see both of these log warnings.

Line 33 will set to Auto
Line 34 will be true and the log on line 36 will be output
Line 40 is also true and the log on line 42 will be output

Reply:
The old definition will have the default passed in as auto, hence the second warning will not be shown as it excludes that.

Reply:
You are right! My bad.